packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/EnumTypeProvider.cs

```diff
@@ -0,0 +1,363 @@
+
+namespace Microsoft.Generator.CSharp
+{
+    public class EnumTypeProvider : TypeProvider
+    {
+        private readonly IReadOnlyList<InputEnumTypeValue> _allowedValues;
+        private readonly ModelTypeMapping? _typeMapping;
+
+        public EnumTypeProvider(InputEnumType input, SourceInputModel? sourceInputModel) : base(sourceInputModel)
+        {
+            _allowedValues = input.AllowedValues;
+            _deprecated = input.Deprecated;
+
+            if (input.Accessibility == "internal")
+            {
+                DeclarationModifiers = TypeSignatureModifiers.Internal;
+            }
+            else
+            {
+                DeclarationModifiers = TypeSignatureModifiers.Public;
+            }
+
+            var isExtensible = input.IsExtensible;
+            if (ExistingType != null)
+            {
+                isExtensible = ExistingType.TypeKind switch
+                {
+                    TypeKind.Enum => false,
+                    TypeKind.Struct => true,
+                    _ => throw new InvalidOperationException(
+                        $"{ExistingType.ToDisplayString()} cannot be mapped to enum," +
+                        $" expected enum or struct got {ExistingType.TypeKind}")
+                };
+
+                _typeMapping = sourceInputModel?.CreateForModel(ExistingType);
+            }
+
+            if (isExtensible)
+            {
+                // extensible enums are implemented by readonly structs
+                DeclarationModifiers |= TypeSignatureModifiers.Partial | TypeSignatureModifiers.ReadOnly;
+            }
+
+            Name = input.Name.ToCleanName();
+            Namespace = GetDefaultModelNamespace(CodeModelPlugin.Instance.Configuration.Namespace);
+            IsExtensible = isExtensible;
+            ValueType = CodeModelPlugin.Instance.TypeFactory.CreateCSharpType(input.EnumValueType);
+            IsStringValueType = ValueType.Equals(typeof(string));
+            IsIntValueType = ValueType.Equals(typeof(int)) || ValueType.Equals(typeof(long));
+            IsFloatValueType = ValueType.Equals(typeof(float)) || ValueType.Equals(typeof(double));
+            IsNumericValueType = IsIntValueType || IsFloatValueType;
+
+            Description = input.Description;
+
+            _valueField = new FieldDeclaration(FieldModifiers.Private | FieldModifiers.ReadOnly, ValueType, "_value");
+
+            if (!IsExtensible)
+            {
+                SerializationProviders = CodeModelPlugin.Instance.GetSerializationTypeProviders(this);
+            }
+        }
+
+        private readonly FieldDeclaration _valueField;
+
+        public CSharpType ValueType { get; }
+        public bool IsExtensible { get; }
+        public bool IsIntValueType { get; }
+        internal bool IsFloatValueType { get; }
+        public bool IsStringValueType { get; }
+        internal bool IsNumericValueType { get; }
+        public string? Description { get; }
```

Comment:
is this the xml summar above the class / struct / enum line?  If so shouldn't this live in the declaration at the TypeProvider level?

Reply:
Yeah this is exactly the thing.
Unfortunately `TypeProvider` does not have it yet.
I will open another PR to do that and paste the link later here

Reply:
I extracted a few things from this PR that are off-topic into a new PR: https://github.com/microsoft/typespec/pull/3431